elasticgraph-warehouse/lib/elastic_graph/warehouse/schema_definition/factory_extension.rb

```diff
@@ -81,11 +84,39 @@ def new_scalar_type(name)
         def new_union_type(name)
           super(name) do |type|
             type.extend ObjectInterfaceAndUnionExtension
+            extend_index_method(type)
             # :nocov: -- currently all invocations have a block
             yield type if block_given?
             # :nocov:
           end
         end
+
+        private
+
+        # Wraps the index method on the type to extend created indices with IndexExtension.
+        #
+        # @param type [Object] the type to extend
+        # @return [void]
+        def extend_index_method(type)
+          type.singleton_class.prepend(IndexMethodWrapper)
+        end
```

Comment:
While this is a clever way to hermetically extend the `index` API I think we can simplify it if we just offer a `new_index` hook from the `Factory` API so that you can hook into it in the same way you do all the other `new_*` methods.

To that end, I've prepared #953. Care to review it and then do you want to rebase this on top to leverage that?